server/src/main/java/org/elasticsearch/common/logging/DeprecationIndexingService.java

```diff
@@ -0,0 +1,154 @@
+import org.elasticsearch.action.index.IndexRequestBuilder;
+import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.OriginSettingClient;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.metadata.ComposableIndexTemplate;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.elasticsearch.common.Strings.isNullOrEmpty;
+
+/**
+ * This service is responsible for writing deprecation messages to a data stream. It also creates
+ * the data stream if necessary. The writing of messages can be toggled using the
+ * {@link #WRITE_DEPRECATION_LOGS_TO_INDEX} setting.
+ */
+public class DeprecationIndexingService implements ClusterStateListener {
+    private static final Logger LOGGER = LogManager.getLogger(DeprecationIndexingService.class);
+
+    private static final String DATA_STREAM_NAME = "logs-deprecation-elasticsearch";
+    private static final String TEMPLATE_NAME = DATA_STREAM_NAME + "-template";
+    private static final String TEMPLATE_MAPPING = TEMPLATE_NAME + ".json";
+
+    private static final String DEPRECATION_ORIGIN = "deprecation";
+
+    public static final Setting<Boolean> WRITE_DEPRECATION_LOGS_TO_INDEX = Setting.boolSetting(
+        "cluster.deprecation_indexing.enabled",
+        true,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    private final Client client;
+    private boolean hasTriedToLoadTemplate = false;
+    private volatile boolean isEnabled = false;
+
+    public DeprecationIndexingService(ClusterService clusterService, Client client) {
+        this.client = new OriginSettingClient(client, DEPRECATION_ORIGIN);
+
+        clusterService.addListener(this);
+    }
+
+    /**
+     * Indexes a deprecation message.
+     * @param key       the key that was used to determine if this deprecation should have been be logged.
+     *                  Useful when aggregating the recorded messages.
+     * @param message   the message to log
+     * @param xOpaqueId the associated "X-Opaque-ID" header value if any, or <code>null</code>
+     * @param params    parameters to the message, if any
+     */
+    public void writeMessage(String key, String message, String xOpaqueId, Object[] params) {
+        if (this.isEnabled == false) {
+            return;
+        }
+
+        Map<String, Object> payload = new HashMap<>();
```

Comment:
we discussed this offline, but I will leave the comment for further discussion. Should we unify the format/fields generated by deprecation logs with what will be indexed?
Should we expect that ingesting 'manually' deprecation logs would give the same result (same format/fields in deprecation index) as when indexing from here?

I wonder if using the layout we use for formatting deprecation logs to ECS
```
SLogMessage deprecationMessage = DeprecatedMessage.of("x", "ss {}", 1);
LogEvent event = Log4jLogEvent.newBuilder()
    .setMessage(deprecationMessage)
    .setLoggerName("logger_name")
    .build();
final LoggerContext context = (LoggerContext) LogManager.getContext(false);
Layout<? extends Serializable> deprecated = context.getConfiguration().getAppender("deprecated").getLayout();
deprecated.toSerializable(event)
```

I am happy with simplest first approach, but might be worth considering this for the future